File: src/main/java/io/colyseus/Room.java
package io.colyseus;

import io.colyseus.serializer.SchemaSerializer;
import io.colyseus.serializer.schema.Schema;
import org.java_websocket.framing.CloseFrame;

import java.lang.reflect.InvocationTargetException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Map;

public class Room<T extends Schema> {

    public abstract static class Listener<T extends Schema> {

        protected Listener() {

        }

        /**
         * This event is triggered when the client leave the room.
         */
        protected void onLeave() {

        }

        /**
         * This event is triggered when some error occurs in the room handler.
         */
        protected void onError(Exception e) {

        }

        /**
         * This event is triggered when the server sends a message directly to the client.
         */
        protected void onMessage(Object message) {

        }

        /**
         * This event is triggered when the client successfuly joins the room.
         */
        protected void onJoin() {

        }

        /**
         * This event is triggered when the server updates its state.
         */
        protected void onStateChange(T state, boolean isFirstState) {

        }
    }

    public T state;

    private Class<T> stateType;

    // private LinkedHashMap<String, Object> options;

    /**
     * The unique identifier of the room.
     */
    private String id;

    /**
     * Unique session identifier.
     */
    private String sessionId;

    /**
     * Name of the room handler. Ex: "battle".
     */

    private String name;
    private Listener listener;
    private Connection connection;
    private byte[] _previousState;
    // private ObjectMapper msgpackMapper;
    private SchemaSerializer<T> serializer;
    private int previousCode;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getSessionId() {
        return sessionId;
    }

    void setSessionId(String sessionId) {
        this.sessionId = sessionId;
    }

    public String getId() {
        return id;
    }

    void setId(String id) {
        this.id = id;
    }

    /*LinkedHashMap<String, Object> getOptions() {
        return options;
    }

    public void setOptions(LinkedHashMap<String, Object> options) {
        this.options = options;
    }*/

//    public LinkedHashMap<String, Object> getState() {
//        return state;
//    }

    Room(Class<T> type, String roomName) {
//        super(new LinkedHashMap<String, Object>());
//        System.out.println("Room created: name: " + roomName + ", options: " + options);
        this.stateType = type;
        this.name = roomName;
        // this.options = options;
        // this.msgpackMapper = new ObjectMapper(new MessagePackFactory());
        try {
            serializer = new SchemaSerializer<>(stateType);
            state = serializer.state;
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    void connect(String endpoint, Map<String, String> httpHeaders) throws URISyntaxException {
//        System.out.println("Room is connecting to " + endpoint);
        this.connection = new Connection(new URI(endpoint), httpHeaders, new Connection.Listener() {
            @Override
            public void onError(Exception e) {
                //System.err.println("Possible causes: room's onAuth() failed or maxClients has been reached.");
                if (listener != null) listener.onError(e);
            }

            @Override
            public void onClose(int code, String reason, boolean remote) {
                if (code == CloseFrame.PROTOCOL_ERROR && reason != null && reason.startsWith("Invalid status code received: 401")) {
                    if (listener != null) listener.onError(new Exception(reason));
                }
                if (listener != null) listener.onLeave();
                //client.onRoomLeave(id);
//                removeAllListeners();
            }

            @Override
            public void onOpen() {

            }

            @Override
            public void onMessage(ByteBuffer buf) {
                Room.this.onMessageCallback(buf);
            }
        });
    }

    private void onMessageCallback(ByteBuffer buf) {
        try {
            if (previousCode == 0) {
                byte code = buf.get();

                if (code == Protocol.JOIN_ROOM) {
                    byte[] bytes = new byte[buf.get()];
                    buf.get(bytes, 0, bytes.length);
                    String serializerId = new String(bytes, StandardCharsets.UTF_8);
                    if (serializerId.equals("fossil-delta")) {
                        throw new Error("fossil-delta is not supported");
                    }
                    if (buf.hasRemaining()) {
                        byte[] b = new byte[buf.remaining()];
                        buf.get(b, 0, b.length);
                        serializer.handshake(b);
                    }
                    if (listener != null) listener.onJoin();
                } else if (code == Protocol.JOIN_ERROR) {
                    int length = buf.get();
                    byte[] bytes = new byte[length];
                    buf.get(bytes, 0, length);
                    String message = new String(bytes, StandardCharsets.UTF_8);
                    if (listener != null) listener.onError(new Exception(message));
                } else if (code == Protocol.LEAVE_ROOM) {
                    leave();
                } else {
                    previousCode = code;
                }

            } else {
                if (buf.hasRemaining()) {
                    if (previousCode == Protocol.ROOM_STATE) {
                        byte[] bytes = new byte[buf.remaining()];
                        buf.get(bytes);
                        setState(bytes);
                    } else if (previousCode == Protocol.ROOM_STATE_PATCH) {
                        byte[] bytes = new byte[buf.remaining()];
                        buf.get(bytes);
                        patch(bytes);
                    } else if (previousCode == Protocol.ROOM_DATA) {
                        byte[] bytes = new byte[buf.remaining()];
                        //buf.get(bytes);
                        System.out.println("ROOM_DATA: " + new String(bytes, StandardCharsets.UTF_8));
//                        Object data = msgpackMapper.readValue(bytes, new TypeReference<Object>() {
//                        });
//                        for (Listener listener : listeners) {
//                            if (listener != null) listener.onMessage(data);
//                        }
                    }
                }
                previousCode = 0;
            }
        } catch (Exception e) {
            if (listener != null) listener.onError(e);
        }
    }

    private void dispatchOnMessage(Object message) {
        if (listener != null) listener.onMessage(message);
    }

//    public void removeAllListeners() {
//        this.listeners.clear();
//    }

    /**
     * Disconnect from the room.
     */
    public void leave() {
        leave(true);
    }

    public void leave(boolean consented) {
        if (id != null) {
            if (consented) {
                connection.send(Protocol.LEAVE_ROOM);
            } else {
                connection.close();
            }

        } else {
            if (listener != null) listener.onLeave();
        }
    }

    /**
     * Send message to the room handler.
     */
    public void send(Object data) {
        if (this.connection != null)
            this.connection.send(Protocol.ROOM_DATA, data);
        else {
            // room is created but not joined yet
            if (listener != null) listener.onError(new Exception("send error: Room is created but not joined yet"));
        }
    }


    public boolean hasJoined() {
        return this.sessionId != null;
    }

    private void setState(byte[] encodedState) throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException, NoSuchFieldException {
        serializer.setState(encodedState);
        if (listener != null) listener.onStateChange(serializer.state._clone(), true);
    }

    private void patch(byte[] delta) throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException, NoSuchFieldException {
        serializer.patch(delta);
        if (listener != null) listener.onStateChange(serializer.state, false);
    }
}

File: src/main/java/io/colyseus/Room.kt
package io.colyseus

import com.fasterxml.jackson.databind.ObjectMapper
import io.colyseus.serializer.SchemaSerializer
import io.colyseus.serializer.schema.*
import io.colyseus.serializer.schema.Iterator
import org.java_websocket.framing.CloseFrame
import org.msgpack.jackson.dataformat.MessagePackFactory
import java.net.URI
import java.net.URISyntaxException
import java.nio.ByteBuffer

class Room<T : Schema> internal constructor(schema: Class<T>, var name: String) {

    public var onLeave: ((code: Int) -> Unit)? = null
    public var onError: ((code: Int, message: String?) -> Unit)? = null
    public var onJoin: (() -> Unit)? = null
    public var onStateChange: ((state: T, isFirstState: Boolean) -> Unit)? = null
    public val onMessageHandlers = hashMapOf<String, MessageHandler<*>>()

    class MessageHandler<out K> constructor(
            val type: Class<out K>,
            val handler: ((message: Any) -> Unit)?,
    )

    var id: String? = null
    var sessionId: String? = null
    private var connection: Connection? = null

    private val msgpackMapper = ObjectMapper(MessagePackFactory())
    private var serializer = SchemaSerializer(schema)
    public val state: T = serializer.state as T

    @Throws(URISyntaxException::class)
    fun connect(endpoint: String, httpHeaders: Map<String, String>? = null) {
//        System.out.println("Room is connecting to " + endpoint)
        connection = Connection(URI(endpoint), httpHeaders)
        connection?.onError = { e -> onError?.invoke(-1, e.message) }
        connection?.onClose = { code, reason, _ ->
            if (code == CloseFrame.PROTOCOL_ERROR &&
                    reason != null &&
                    reason.startsWith("Invalid status code received: 401")) {
                onError?.invoke(-1, reason)
            }
            onLeave?.invoke(code)
        }
        connection?.onMessage = { buf: ByteBuffer ->
            val length = buf.remaining()
            val bytes = ByteArray(length)
            buf[bytes, 0, length]
            this.parseMessage(bytes)
        }
        connection?.connect()
    }

    private fun parseMessage(bytes: ByteArray) {
        try {
            val code = bytes[0].toInt() and 0xFF
            when (code) {
                Protocol.JOIN_ROOM -> {
                    var offset = 1
                    val serializerId = String(bytes, offset + 1, bytes[offset].toInt() and 0xFF)
                    offset += serializerId.length + 1
                    if (serializerId == "fossil-delta") {
                        throw Error("fossil-delta is not supported")
                    }
                    if (bytes.size > offset) {
                        serializer.handshake(bytes, offset)
                    }
                    onJoin?.invoke()
                    connection?._send(Protocol.JOIN_ROOM)
                }
                Protocol.ERROR -> {
                    val it = Iterator(1)
                    val errorCode: Int = Decoder.decodeNumber(bytes, it).toInt()
                    val errorMessage = Decoder.decodeString(bytes, it)
                    onError?.invoke(errorCode, errorMessage)
                }
                Protocol.ROOM_DATA_SCHEMA -> {
                    val messageType: Class<*> = Context.instance.get(bytes[1].toInt() and 0xFF) as Class<*>
                    val message = messageType.getConstructor().newInstance() as Schema
                    message.decode(bytes, Iterator(2))
                    val messageHandler = onMessageHandlers["s" + messageType.typeName]
                    if (messageHandler != null) {
                        messageHandler.handler?.invoke(message)
                    } else {
                        println("No handler for type " + messageType.typeName)
                    }
                }
                Protocol.LEAVE_ROOM -> leave()
                Protocol.ROOM_STATE -> setState(bytes, 1)
                Protocol.ROOM_STATE_PATCH -> patch(bytes, 1)
                Protocol.ROOM_DATA -> {
                    val messageHandler: MessageHandler<*>?
                    val type: Any
                    val it = Iterator(1)
                    if (Decoder.numberCheck(bytes, it)) {
                        type = Decoder.decodeNumber(bytes, it).toInt()
                        messageHandler = onMessageHandlers["i$type"]
                    } else {
                        type = Decoder.decodeString(bytes, it)
                        messageHandler = onMessageHandlers[type.toString()]
                    }
                    if (messageHandler != null) {
                        if (bytes.size > it.offset) {
                            messageHandler.handler?.invoke(
                                    msgpackMapper.readValue(bytes[it.offset..bytes.size], messageHandler.type)!!
                            )
                        }
                    } else {
                        println("No handler for type $type" + if (type is Int) " (${Context.instance.get(type)?.simpleName})" else "")
                    }
                }
            }
        } catch (e: Exception) {
            leave(false)
            e.printStackTrace()
            onError?.invoke(-1, e.message)
        }
    }

    // Disconnect from the room.
    @JvmOverloads
    fun leave(consented: Boolean = true) {
        if (connection != null) {
            if (consented) {
                connection!!._send(Protocol.LEAVE_ROOM)
            } else {
                connection!!.close()
            }
        } else {
            onLeave?.invoke(4000) // "consented" code
        }
    }

    // Send a message by number type, without payload
    public fun send(type: Int) {
        connection?.send(byteArrayOfInts(Protocol.ROOM_DATA, type))
    }

    // Send a message by number type with payload
    public fun send(type: Int, message: Any) {
        val initialBytes: ByteArray = byteArrayOfInts(Protocol.ROOM_DATA, type)
        val encodedMessage: ByteArray = msgpackMapper.writeValueAsBytes(message)
        connection?.send(initialBytes + encodedMessage)
    }

    // Send a message by string type, without payload
    public fun send(type: String) {
        val encodedType: ByteArray = type.toByteArray()
        val initialBytes: ByteArray = Encoder.getInitialBytesFromEncodedType(encodedType)
        connection?.send(initialBytes + encodedType)
    }

    // Send a message by string type with payload
    public fun send(type: String, message: Any) {
        val encodedMessage: ByteArray = msgpackMapper.writeValueAsBytes(message)
        val encodedType: ByteArray = type.toByteArray()
        val initialBytes: ByteArray = Encoder.getInitialBytesFromEncodedType(encodedType)
        connection?.send(initialBytes + encodedType + encodedMessage)
    }

    public inline fun <reified MessageType> onMessage(
            type: String,
            noinline handler: ((message: MessageType) -> Unit)?,
    ) {
        onMessageHandlers[type] = MessageHandler(
                MessageType::class.java,
                handler as ((Any) -> Unit)?
        )
    }

    public inline fun <reified MessageType> onMessage(
            type: Int,
            noinline handler: ((message: MessageType) -> Unit)?,
    ) {
        onMessageHandlers["i$type"] = MessageHandler(
                MessageType::class.java,
                handler as ((Any) -> Unit)?
        )
    }

    public inline fun <reified MessageType> onMessage(
            noinline handler: ((message: MessageType) -> Unit)?,
    ) {
        onMessageHandlers["s" + MessageType::class.java.name] = MessageHandler(
                MessageType::class.java,
                handler as ((Any) -> Unit)?
        )
    }

    public fun hasJoined(): Boolean {
        return sessionId != null
    }

    @Throws(Exception::class)
    private fun setState(encodedState: ByteArray, offset: Int = 0) {
        serializer.setState(encodedState, offset)
        onStateChange?.invoke(serializer.state as T, true)
    }

    @Throws(Exception::class)
    private fun patch(delta: ByteArray, offset: Int = 0) {
        serializer.patch(delta, offset)
        onStateChange?.invoke(serializer.state as T, false)
    }
}
